sql/core/src/test/scala/org/apache/spark/sql/execution/BaseScriptTransformationSuite.scala

```diff
@@ -0,0 +1,367 @@
+          new Date(2020, 7, 3), new CalendarInterval(7, 3, 3000), Array(6, 7, 8),
+          Map("c" -> 3), new TestUDT.MyDenseVector(Array(1, 2, 3)), new SimpleTuple(1, 1L))
+      ).toDF("a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l")
+        .select('a, 'b, 'c, 'd, 'e, 'f, 'g, 'h, 'i, 'j, 'k, 'l,
+          struct('a, 'b).as("m"), unhex('a).as("n"), lit(true).as("o")
+        ) // Note column d's data type is Decimal(38, 18)
+
+      // Can't support convert script output data to ArrayType/MapType/StructType now,
+      // return these column still as string.
+      // For UserDefinedType, if user defined deserialize method to support convert string
+      // to UserType like [[SimpleTupleUDT]], we can support convert to this UDT, else we
+      // will return null value as column.
+      checkAnswer(
+        df,
+        (child: SparkPlan) => createScriptTransformationExec(
+          input = Seq(
+            df.col("a").expr,
+            df.col("b").expr,
+            df.col("c").expr,
+            df.col("d").expr,
+            df.col("e").expr,
+            df.col("f").expr,
+            df.col("g").expr,
+            df.col("h").expr,
+            df.col("i").expr,
+            df.col("j").expr,
+            df.col("k").expr,
+            df.col("l").expr,
+            df.col("m").expr,
+            df.col("n").expr,
+            df.col("o").expr),
+          script = "cat",
+          output = Seq(
+            AttributeReference("a", IntegerType)(),
+            AttributeReference("b", StringType)(),
+            AttributeReference("c", DoubleType)(),
+            AttributeReference("d", ByteType)(),
+            AttributeReference("e", DecimalType(38, 18))(),
+            AttributeReference("f", TimestampType)(),
+            AttributeReference("g", DateType)(),
+            AttributeReference("h", CalendarIntervalType)(),
+            AttributeReference("i", StringType)(),
+            AttributeReference("j", StringType)(),
+            AttributeReference("k", StringType)(),
+            AttributeReference("l", new SimpleTupleUDT)(),
+            AttributeReference("m", StringType)(),
+            AttributeReference("n", BinaryType)(),
+            AttributeReference("o", BooleanType)()),
+          child = child,
+          ioschema = defaultIOSchema
+        ),
+        df.select(
+          'a, 'b, 'c, 'd, 'e, 'f, 'g, 'h,
+          'i.cast("string"),
+          'j.cast("string"),
+          'k.cast("string"),
+          'l, 'm.cast("string"), 'n, 'o).collect())
+    }
+  }
+
+  test("SPARK-32106: TRANSFORM should return null when return string incompatible") {
+    checkAnswer(
+      sql(
+        """
+          |SELECT TRANSFORM(a, b, c)
+          |USING 'cat' as (a int, b int , c int)
+          |FROM (
+          |SELECT
+          |1 AS a,
+          |"a" AS b,
+          |CAST(2000 AS timestamp) AS c
+          |) tmp
+        """.stripMargin),
+      identity,
+      Row(1, null, null) :: Nil)
+  }
```

Comment:
This part is duplicated with `SQLQueryTestSuite`, so we can remove it. Instead, I think it would better to add more tests for all the incompatible patterns (`string`->short, `string`->long, ...) for better test coverage. Could you?

Reply:
> This part is duplicated with `SQLQueryTestSuite`, so we can remove it. Instead, I think it would better to add more tests for all the incompatible patterns (`string`->short, `string`->long, ...) for better test coverage. Could you?

Done